Improve `Tensor` type hints

henribru opened the pull request:
I've typed some attributes from https://github.com/pytorch/pytorch/blob/ee920b92c4d52c3481ce4fa1594893fde8ccdd93/torch/csrc/autograd/python_variable.cpp#L490 that were not included in the stubs so that MyPy will be aware of them. I made sure to only add those attributes that are mentioned somewhere in the documentation. If there are attributes mentioned in the documentation that are not meant to be part of the public API (or the opposite), please let me know. I've also made sure that attributes that can't be set are typed as read-only properties. If setting `dtype`, `shape`, `device` or `names` directly is not part of the public API, let me know and I'll make them properties as well.

I've also added `__len__`, `__iter__` and `__contains__`, which means MyPy will no longer complain about `len(t)`, `t1 in t2` and `for t1 in t2`.

Shameless plug: I have another typing-related PR here that needs review: https://github.com/pytorch/pytorch/pull/27445

Fixes #28457

henribru committed:
Improve Tensor type hints

- Add some missing attributes
- Add some missing methods
- Type read-only attributes as properties

Fixes #28457

fmassa reviewed (approved):
LGTM, thanks!

fmassa commented on torch/__init__.pyi.in:
```diff
@@ -83,16 +83,31 @@ class Generator:
+    @property
+    def ndim(self) -> _int: ...
+    @property
+    def layout(self) -> layout:
```
This is an invalid syntax:
```
Oct 24 12:22:38 test_doc_examples (__main__.TestTypeHints) ... torch/__init__.pyi:105: error: invalid syntax
```

henribru replied:
Sorry about that, should be fixed now.

fmassa reviewed (changes requested):
Test failures seems related https://circleci.com/gh/pytorch/pytorch/3336395?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

henribru committed:
Add missing ellipsis

henribru committed:
Work around possible MyPy bug

I suspect that https://github.com/python/mypy/issues/4146 could
cause an issue with the layout property.

henribru commented:
https://github.com/python/mypy/issues/4146 is causing problems it seems. I'll have a look.

henribru committed:
Avoid shadowing method names in types

Workaround for https://github.com/python/mypy/issues/4146
